types/src/v0/impls/stake_table.rs

```diff
@@ -26,84 +37,217 @@ use hotshot_types::{
     },
     PeerConfig,
 };
+use indexmap::IndexMap;
 use itertools::Itertools;
 use thiserror::Error;
 
 use super::{
     traits::StateCatchup,
-    v0_3::{DAMembers, StakeTable, StakeTables},
+    v0_3::{DAMembers, Delegator, StakeTable, Validator},
     Header, L1Client, Leaf2, NodeState, PubKey, SeqTypes,
 };
 
 type Epoch = <SeqTypes as NodeType>::Epoch;
 
-impl StakeTables {
-    pub fn new(stake_table: StakeTable, da_members: DAMembers) -> Self {
-        Self {
-            stake_table,
-            da_members,
+/// Create the consensus and DA stake tables from L1 events
+///
+/// This is a pure function, to make it easily testable.
+///
+/// We expect have at most a few hundred EVM events in the
+/// PermissionedStakeTable contract over the liftetime of the contract so it
+/// should not significantly affect performance to fetch all events and
+/// perform the computation in this functions once per epoch.
+pub fn from_l1_events(
+    registered: Vec<(ValidatorRegistered, Log)>,
+    deregistered: Vec<(ValidatorExit, Log)>,
+    delegated: Vec<(Delegated, Log)>,
+    undelegated: Vec<(Undelegated, Log)>,
+    keys_update: Vec<(ConsensusKeysUpdated, Log)>,
+) -> IndexMap<Address, Validator<BLSPubKey>> {
+    // TODO: return RESULT
+    // TODO: Handle consensus keys update event
+    let mut st_map = BTreeMap::new();
+    let mut bls_keys = HashSet::new();
+    let mut schnor_keys = HashSet::new();
+    for (reg, log) in registered {
+        let bls = bls_alloy_to_jf2(reg.blsVk.clone());
+        let state_ver_key = edward_bn254point_to_state_ver(reg.schnorrVk.clone());
+        if bls_keys.contains(&bls) {
+            tracing::error!("bls key {bls:?} already used");
+            continue;
+        }
+
+        if schnor_keys.contains(&state_ver_key) {
+            tracing::error!("schnor verifying key {bls:?} already used");
+            continue;
         }
+        st_map.insert(
+            (
+                log.block_number.unwrap(),
+                log.transaction_index.unwrap(),
```

Comment:
the log index should be the index in the block, so I think we don't need the tx index?

https://docs.rs/alloy/0.12.5/alloy/rpc/types/struct.Log.html#structfield.log_index